File: src/main/java/com/extendedae_plus/common/item/ItemProviderController.java
package com.extendedae_plus.common.item;

import appeng.api.networking.IInWorldGridNodeHost;
import com.extendedae_plus.common.init.ModItems;
import com.extendedae_plus.common.menu.MenuProviderController;
import com.extendedae_plus.util.UtilGetKey;
import net.minecraft.network.chat.Component;
import net.minecraft.world.InteractionResult;
import net.minecraft.world.MenuProvider;
import net.minecraft.world.entity.player.Inventory;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.inventory.AbstractContainerMenu;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.context.UseOnContext;
import org.jetbrains.annotations.Nullable;

public class ItemProviderController extends Item implements MenuProvider {
    public ItemProviderController() {
        super(new Properties().stacksTo(1));
    }

    @Override
    public InteractionResult useOn(UseOnContext context) {
        var level = context.getLevel();
        if (level.isClientSide()) return InteractionResult.sidedSuccess(true);

        var item = context.getItemInHand();
        if (!(item.getItem() instanceof MenuProvider provider)) return super.useOn(context);

        var player = context.getPlayer();
        if (player == null) return super.useOn(context);

        var rawBlockEntity = level.getBlockEntity(context.getClickedPos());
        if (rawBlockEntity instanceof IInWorldGridNodeHost) {
            player.openMenu(provider, buf -> {
                buf.writeBlockPos(context.getClickedPos());
                buf.writeUtf(context.getClickedFace().getName());
            });
        }
        return InteractionResult.CONSUME;
    }

    @Override
    public Component getDisplayName() {
        return new UtilGetKey(UtilGetKey.screen)
                .item(ModItems.PROVIDER_CONTROLLER)
                .build();
    }

    @Override
    public @Nullable AbstractContainerMenu createMenu(int i, Inventory inventory, Player player) {
        return new MenuProviderController(i, inventory, null);
    }
}

File: src/main/kotlin/com/extendedae_plus/common/registry/item/ItemProviderController.kt
package com.extendedae_plus.common.registry.item

import appeng.api.networking.IInWorldGridNodeHost
import com.extendedae_plus.common.init.EAEPItems
import com.extendedae_plus.network.CPacketProviderControllerOperation
import com.extendedae_plus.network.CPacketProviderControllerOperation.Operation
import com.extendedae_plus.util.UtilKeyBuilder
import com.extendedae_plus.util.UtilTextComponent
import com.fish.fishlib.network.base.PacketGeneric.Companion.sendToServer
import com.fish.fishlib.util.keyBuilder.Patterns
import com.fish.fishlib.util.keyBuilder.containerComponent
import com.fish.fishlib.util.keyBuilder.newContainerComponent
import net.minecraft.ChatFormatting
import net.minecraft.network.chat.Component
import net.minecraft.network.chat.HoverEvent
import net.minecraft.network.chat.MutableComponent
import net.minecraft.world.InteractionResult
import net.minecraft.world.item.Item
import net.minecraft.world.item.ItemStack
import net.minecraft.world.item.context.UseOnContext
import net.neoforged.api.distmarker.Dist
import net.neoforged.api.distmarker.OnlyIn

class ItemProviderController : Item(Properties().stacksTo(1)) {
    override fun onItemUseFirst(stack: ItemStack, context: UseOnContext): InteractionResult {
        val level = context.level
        if (!level.isClientSide()) return InteractionResult.sidedSuccess(false)

        val player = context.player ?: return super.useOn(context)

        val tileRaw = level.getBlockEntity(context.clickedPos)
        if (tileRaw is IInWorldGridNodeHost) this.cui(stack, context)?.let {
            player.displayClientMessage(it, false)
        }
        return InteractionResult.SUCCESS
    }

    // 我们有最好的函数式cui构造😋
    @OnlyIn(Dist.CLIENT)
    private fun cui(stack: ItemStack, context: UseOnContext) = UtilKeyBuilder.of(Patterns.Message)
        .item(EAEPItems.ControllerProvider)
        .addStr("cui")
        .newContainerComponent()
        .args(stack.displayName)
        .buildInto("title") {
            it.withStyle {
                it.withHoverEvent(
                    HoverEvent(
                        HoverEvent.Action.SHOW_TEXT,
                        UtilKeyBuilder.of(Patterns.Message)
                            .item(EAEPItems.ControllerProvider)
                            .addStr("cui")
                            .addStr("title")
                            .addStr("tip")
                            .build()
                    )
                )
            }
        }
        .newLine()
        .section("button") {
            it
                .section("blocking") {
                    it
                        .buildInto("normal", context.send(blockingNormal = Operation.TOGGLE))
                        .buildInto("smart", context.send(blockingSmart = Operation.TOGGLE))
                }
                .buildInto("doubling", context.send(doubling = Operation.TOGGLE))
                .newLine()
                .section("of_all") {
                    it
                        .buildInto(
                            "on", context.send(
                                Operation.SET_TRUE,
                                Operation.SET_TRUE,
                                Operation.SET_TRUE,
                                ChatFormatting.GOLD
                            )
                        ).buildInto(
                            "off", context.send(
                                Operation.SET_FALSE,
                                Operation.SET_FALSE,
                                Operation.SET_FALSE,
                                ChatFormatting.GOLD
                            )
                        )
                }
        }
        .newLine()
        .append(Component.literal("=".repeat(8 * 2 + stack.displayName.string.length)))
//        .buildInto("end") {
//            it.withStyle {
//                it.withHoverEvent(
//                    HoverEvent(
//                        HoverEvent.Action.SHOW_TEXT,
//                        UtilKeyBuilder.of(EAEPItems.ControllerProvider)
//                            .addStr("cui")
//                            .addStr("end")
//                            .addStr("tip")
//                            .build()
//                    )
//                )
//            }
//        }
        .containerComponent

    @OnlyIn(Dist.CLIENT)
    private fun UseOnContext.send(
        blockingNormal: Operation = Operation.NOOP,
        blockingSmart: Operation = Operation.NOOP,
        doubling: Operation = Operation.NOOP,
        color: ChatFormatting = ChatFormatting.GREEN
    ) = { component: MutableComponent ->
        component.withStyle {
            it.withClickEvent(UtilTextComponent.ClickEventCustomizable {
                CPacketProviderControllerOperation(
                    blockingNormal,
                    blockingSmart,
                    doubling,
                    this.clickedPos,
                    this.clickedFace
                ).sendToServer()
            }).withColor(color)
        }
    }
}
